### DISCUSSION TITLE
Polymer Solidification Simulation from Wang et al., 2013

### DISCUSSION THREAD

**Discussion by Josh-Web-700** [idaholab/moose#28036] [Q&A Modules: Phase field]
Hello MOOSE experts,

I am working on reproducing the solidification simulations for polymer crystallization from the article, “[ A Phase-Field Model for Simulating Various Spherulite Morphologies of Semi-Crystalline Polymers](https://iopscience.iop.org/article/10.1088/1674-1056/22/10/106103),” (Wang Xiao-Dong et al 2013 Chinese Phys. B 22 106103). So far, I have not been able to reproduce the crystallographic branching that would be expected in the simulation. I am not sure if this is an error in my understanding of the article’s parameters or in my MOOSE implementation, and any insight would be much appreciated.

### Governing Equations & Variables

There are two governing equations:  a nonconserved Ginzburg-Landau (Allen-Cahn) equation for crystallinity, and a heat equation for temperature. The two equations are shown below in nondimensional form:

![image](https://github.com/idaholab/moose/assets/159085909/71f5f6c7-c4c8-442e-b9f8-14d8819a587f)

Listed below are the expressions used to evaluate the governing equations:
![image](https://github.com/idaholab/moose/assets/159085909/e5e5ca5b-5cc7-4add-bdba-0a2e918e102c)

(below is a modified evaluation for zeta, ζ-overbar)
![image](https://github.com/idaholab/moose/assets/159085909/8244d9b4-7194-42a1-b087-585a82a209d9)
![image](https://github.com/idaholab/moose/assets/159085909/f9bd5199-b2f0-4b2b-a235-58672829d7a8)
![image](https://github.com/idaholab/moose/assets/159085909/bfa25e10-3baa-41aa-a343-a9a4421047e2)

### Nondimensional Parameters

As mentioned, the simulation was executed with nondimensional parameters. The actual parameters at a temperature of 210 degrees Celsius are:

![image](https://github.com/idaholab/moose/assets/159085909/0f0acc5b-4c55-41d5-a309-eeb7b1881235)

The nondimensional parameters were obtained by:

![image](https://github.com/idaholab/moose/assets/159085909/cb484c8f-d4a7-43fd-9d79-71c98748cfd3)

Also, temperature was nondimensionalized as U=(T-T_m)/(T_m-T_c).

If I understood the article correctly, these are the nondimensional parameters that I used:
- zeta_0 = 0.96
- alpha = 0.658
- K =k = 2.07
- kappa_0^2 = 1.2
- W = 10.69
- T_m, T_c, and T_m_0 remain unchanged

###  Numerical Methods and Initial Conditions

According to the authors, these equations were solved on a 512x512 grid where dτ = 0.1 and dx = dy = 1.0. They used homogeneous Neumann boundary conditions (∂ψ/∂τ=0, ∂U/∂τ=0). The initial conditions for the order parameter (ψ) where 0 over the domain, except at the nuclei/seeds, where the order parameter is set equal to zeta_0 (ψ = ζ_0 = 0.96). The initial conditions for temperature (U) are -1 over the domain.

As far as I am aware, the article made no mention of including any thermal noise. I am unsure of this is important, and I assume that I would get at least some branching in my results from the anisotropy function if I was implementing this correctly.

### Expected Results

At a crystallization temperature of 210 degrees Celsius, with j=6 and ε=0.06, the simulation should appear as the following at dimensionless time of 300:

![image](https://github.com/idaholab/moose/assets/159085909/aa8be554-5e82-4b6e-be55-f4a0e51353ef)

I performed the simulation in two ways:  First, with a constant zeta (ζ), taken as ζ= 0.167. Second, I used an **ADDerivativeParsedMaterial** for ζ to be computed in real time.

When I used a constant ζ, I obtained the following result, shown at a nondimensional time of τ = 25.8:

![image](https://github.com/idaholab/moose/assets/159085909/4715bbd8-0617-4490-b0b3-78858b1a0459)

When I used a nonconstant ζ, the simulation had difficulty making it past a nondimensional time of about τ = 2.0.

### MOOSE Program

I followed the format of the MOOSE example program, "ad_snow.i" ([moose/modules/phase_field/examples/anisotropic_interfaces/ad_snow.i](https://github.com/idaholab/moose/blob/next/modules/phase_field/examples/anisotropic_interfaces/ad_snow.i)), since the simulations in that example are fairly similar. I used two custom kernels by adding a scalar coefficient (kappa_0) to the MOOSE kernels **ADACInterfaceKobayashi1** and **ADACInterfaceKobayashi1**.

### **pfm.i**
In case it is helpful, this is the program I used for a nonconstant (temperature-dependent) zeta (ζ). As mentioned, I used a smaller grid to speed up the simulation and also added adaptive mesh refinement from the ad_snow.i template.
```
[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 32
  ny = 32
  xmin = -128
  xmax = 128
  ymin = -128
  ymax = 128
[]

[Variables]
  [psi]
  []
  [U]
  []
[]

[ICs]
  [psiIC]
    type = SmoothCircleIC
    variable = psi
    x1 = 0
    y1 = 0
    radius = 2
    outvalue = 0
    invalue = 0.96
  []
  [UIC]
    type = ConstantIC
    variable = U
    value = -1
  []
[]

[Kernels]
  [psi_dot]
    type = ADTimeDerivative
    variable = psi
  []
  [anisoACinterface1]
    type = ADCoefACInterfaceKobayashi1
    variable = psi
    mob_name = M
  []
  [anisoACinterface2]
    type = ADCoefACInterfaceKobayashi2
    variable = psi
    mob_name = M
  []
  [ACbulk]
    type = ADAllenCahn
    variable = psi
    mob_name = M
    f_name = floc
  []

  [U_dot]
    type = ADTimeDerivative
    variable = U
  []
  [coefDiffusion] 
    type = ADHeatConduction
    variable = U
    thermal_conductivity = 0.658
  []
  [psi_dot_term]
    type = ADCoefCoupledTimeDerivative
    variable = U
    v = psi
    coef = -2.07
  []
[]

[Materials]
  [consts]
    type = ADGenericConstantMaterial
    prop_names  = 'M kappa0Squared'
    prop_values = '1 1.2'
  []
  [local_free_energy]
    type = ADDerivativeParsedMaterial
    property_name = floc
    coupled_variables = 'psi U'
    constant_names = 'W zeta0'
    constant_expressions = '10.69 0.96'
    material_property_names = 'zeta'
    expression = 'W * ( 1/4*psi^4 - (zeta+zeta0)/3*psi^3 + (zeta*zeta0)/2*psi^2)'
    derivative_order = 2
    outputs = exodus
  []
  [./energy_barrier]
    type = ADDerivativeParsedMaterial
    property_name = zeta
    coupled_variables = 'U'
    constant_names = 'pi T_m T_m0 T_c k zeta0'
    constant_expressions = '4*atan(1) 233 242 210 2.07 0.96'
    expression = 'T:=T_m+U*(T_m-T_c); psiHat:=zeta0*(T_m0-T_m)/(T_m0-T); (1+2/pi*atan(k*U)) * (4*zeta0*psiHat - 3*psiHat^2)/(6*zeta0 - 4*psiHat)'
    derivative_order = 2
    outputs = exodus
  [../]
  [anisotropy]
    type = ADInterfaceOrientationMaterial
    op = psi
    anisotropy_strength = 0.06
    eps_bar = 1
    mode_number = 6
    reference_angle = 0
    outputs = exodus
  []
[]

[Executioner]
  type = Transient
  scheme = bdf2
  solve_type = NEWTON

  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu   '

  nl_abs_tol = 1e-10
  nl_rel_tol = 1e-08
  l_max_its = 30
  nl_max_its = 30

  end_time = 300

  [./TimeStepper]
    type = IterationAdaptiveDT
    optimal_iterations = 10
    iteration_window = 2
    dt = 0.1
    growth_factor = 1.1
    cutback_factor = 0.95
  [../]
  [./Adaptivity]
    initial_adaptivity = 2 # Number of times mesh is adapted to initial condition
    refine_fraction = 0.7 # Fraction of high error that will be refined
    coarsen_fraction = 0.1 # Fraction of low error that will coarsened
    max_h_level = 5 # Max number of refinements used, starting from initial mesh (before uniform refinement)
    weight_names = 'psi U'
    weight_values = '1 0.5'
  [../]
[]

[Outputs]
  time_step_interval = 5
  exodus = true
[]
```

### **First Modified Kernel - ADCoefACInterfaceKobayashi1.C**
```
#include "ADCoefACInterfaceKobayashi1.h"

registerMooseObject("pfmsolid2App", ADCoefACInterfaceKobayashi1);

InputParameters
ADCoefACInterfaceKobayashi1::validParams()
{
  InputParameters params = ADKernelGrad::validParams();
  params.addClassDescription("Anisotropic gradient energy Allen-Cahn Kernel Part 1");
  params.addParam<MaterialPropertyName>("mob_name", "L", "The mobility used with the kernel");
  params.addParam<MaterialPropertyName>("eps_name", "eps", "The anisotropic interface parameter");
  params.addParam<MaterialPropertyName>("deps_name", "deps", "The derivative of the anisotropic interface parameter with respect to angle");
  params.addParam<MaterialPropertyName>("kappa0Squared", "kappa0Squared", "The interfaical energy coefficient");
  return params;
}

ADCoefACInterfaceKobayashi1::ADCoefACInterfaceKobayashi1(const InputParameters & parameters)
  : ADKernelGrad(parameters),
    _mob(getADMaterialProperty<Real>("mob_name")),
    _eps(getADMaterialProperty<Real>("eps_name")),
    _deps(getADMaterialProperty<Real>("deps_name")),
    _kappa0Squared(getADMaterialProperty<Real>("kappa0Squared"))
{
}

ADRealGradient
ADCoefACInterfaceKobayashi1::precomputeQpResidual()
{
  // Set modified gradient vector
  const ADRealGradient v(-_grad_u[_qp](1), _grad_u[_qp](0), 0);

  // Define anisotropic interface residual
  return _kappa0Squared[_qp] * _eps[_qp] * _deps[_qp] * _mob[_qp] * v;
}
```

### **Second Modified Kernel - ADCoefACInterfaceKobayashi2.C**
```
#include "ADCoefACInterfaceKobayashi2.h"

registerMooseObject("pfmsolid2App", ADCoefACInterfaceKobayashi2);

InputParameters
ADCoefACInterfaceKobayashi2::validParams()
{
  InputParameters params = ADKernelGrad::validParams();
  params.addClassDescription("Anisotropic Gradient energy Allen-Cahn Kernel Part 2");
  params.addParam<MaterialPropertyName>("mob_name", "L", "The mobility used with the kernel");
  params.addParam<MaterialPropertyName>("eps_name", "eps", "The anisotropic parameter");
  params.addParam<MaterialPropertyName>("kappa0Squared", "kappa0Squared", "The interfacial energy coefficient");
  return params;
}

ADCoefACInterfaceKobayashi2::ADCoefACInterfaceKobayashi2(const InputParameters & parameters)
  : ADKernelGrad(parameters),
    _mob(getADMaterialProperty<Real>("mob_name")),
    _eps(getADMaterialProperty<Real>("eps_name")),
    _kappa0Squared(getADMaterialProperty<Real>("kappa0Squared"))
{
}

ADRealGradient
ADCoefACInterfaceKobayashi2::precomputeQpResidual()
{
  // Set interfacial part of residual
  return _kappa0Squared[_qp] * _eps[_qp] * _eps[_qp] * _mob[_qp] * _grad_u[_qp];
}
```

**Comment by GiudGiud** [comment]
@laagesen

**Reply by laagesen** [reply]
This is pretty far outside of my recent experience, and dendritic sidebranching is a pretty tricky topic in phase-field modeling. I believe that often thermal noise is not explicitly included in such modeling, but often numerical noise resulting from error in numerical solution of the PDEs is enough to induce sidebranching. However this is highly dependent on parameters you use. You might try increasing the strength of the anisotropy (epsilon) or including thermal noise explicitly. You might have a look at our article https://iopscience.iop.org/article/10.1088/1361-651X/ac46dc
also tagging @SudiptaBiswas if she has anything else to add

**Reply by Josh-Web-700** [reply]
Thank you so much for the feedback. That is helpful!
